test/support/test_iterators.h

```diff
@@ -13,520 +13,1670 @@
-#define _TEST_ITERATORS_H
+#ifndef SUPPORT_TEST_ITERATORS_H
+#define SUPPORT_TEST_ITERATORS_H
 
+#include <cassert>
+#include <concepts>
+#include <cstdint>
 #include <iterator>
-#include <cstddef>
+#include <ranges>
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
 
-#define DELETE_FUNCTION = delete
+#include "double_move_tracker.h"
+#include "test_macros.h"
+#include "type_algorithms.h"
 
+
+// This iterator meets C++20's Cpp17OutputIterator requirements, as described
+// in Table 90 ([output.iterators]).
 template <class It>
-class output_iterator
+class cpp17_output_iterator
 {
     It it_;
+    support::double_move_tracker tracker_;
 
-    template <class U>
-    friend class output_iterator;
-
-  public:
-    typedef ::std::output_iterator_tag iterator_category;
-    typedef void value_type;
-    typedef typename ::std::iterator_traits<It>::difference_type difference_type;
-    typedef typename ::std::iterator_traits<It>::pointer pointer;
-    typedef typename ::std::iterator_traits<It>::reference reference;
+    template <class U> friend class cpp17_output_iterator;
+public:
+    typedef          std::output_iterator_tag                  iterator_category;
+    typedef void                                               value_type;
+    typedef typename std::iterator_traits<It>::difference_type difference_type;
+    typedef It                                                 pointer;
+    typedef typename std::iterator_traits<It>::reference       reference;
 
-    It
-    base() const
-    {
-        return it_;
-    }
+    TEST_CONSTEXPR explicit cpp17_output_iterator(It it) : it_(std::move(it)) {}
 
-    output_iterator() {}
-    explicit output_iterator(It it) : it_(it) {}
     template <class U>
-    output_iterator(const output_iterator<U>& u) : it_(u.it_)
-    {
+    TEST_CONSTEXPR cpp17_output_iterator(const cpp17_output_iterator<U>& u) : it_(u.it_), tracker_(u.tracker_) {}
+
+    template <class U, class = typename std::enable_if<std::is_default_constructible<U>::value>::type>
+    TEST_CONSTEXPR_CXX14 cpp17_output_iterator(cpp17_output_iterator<U>&& u)
+        : it_(std::move(u.it_)), tracker_(std::move(u.tracker_)) {
+      u.it_ = U();
     }
 
-    reference operator*() const { return *it_; }
+    TEST_CONSTEXPR reference operator*() const {return *it_;}
 
-    output_iterator&
-    operator++()
-    {
-        ++it_;
-        return *this;
-    }
-    output_iterator
-    operator++(int)
-    {
-        output_iterator tmp(*this);
-        ++(*this);
-        return tmp;
-    }
+    TEST_CONSTEXPR_CXX14 cpp17_output_iterator& operator++() {++it_; return *this;}
+    TEST_CONSTEXPR_CXX14 cpp17_output_iterator operator++(int) {return cpp17_output_iterator(it_++);}
+
+    friend TEST_CONSTEXPR It base(const cpp17_output_iterator& i) { return i.it_; }
 
     template <class T>
-    void operator,(T const &) DELETE_FUNCTION;
+    void operator,(T const &) = delete;
 };
-
+#if TEST_STD_VER > 14
```

Comment:
Do we need to introduce all of these C++11 and C++14 test macros in our tests code? Or it would be better to remove everything because only C++17 is supported

Reply:
it is not our code - it is utility from LLVM. Probably, it make sense keep it as is to minimize differences and make sync process easier.